design-proposals/klip-24-key-column-semantics-in-queries.md

```diff
@@ -0,0 +1,387 @@
+
+CREATE TABLE OUTPUT AS SELECT I1.V0, I2.V2 FROM INPUT I1 JOIN INPUT I2 ON I1.ID = I2.ID;
+-- fails as key column not in projection.
+```
+
+We propose this is acceptable, and preferable to the current model.
+
+There is no reason why ksqlDB should not support creating sources in the future that do not contain
+the key column. See the [rejected alternatives](#rejected_alternatives) section for more info.
+
+### Allowing the key column to be copied to the value
+
+Hopefully, all of the above seems simple and clear. Now for one of the more tricky and murky bits...
+
+The observant among you may be thinking 'ah, but what if I _want_ to put a copy of the key column
+into the value?'. With current syntax you can do:
+
+```sql
+-- current syntax:
+CREATE TABLE OUTPUT AS SELECT V0, V1, ID AS V2 FROM INPUT;
+-- resulting schema: ID INT KEY, V0 INT, V1 INT, V2 INT
+```
+
+With the proposed syntax the `ID AS V2` would be treated an aliased key column, resulting in the
+schema 'V2 INT KEY, V0 INT, V1 INT'. So, how does the user create an copy of the key in the value,
+if that's what they need to do?
+
+We propose introducing a `COPY` function, (name TBD - maybe 'AS_VALUE'?), that can be used to
+indicate the key column should be copied as a value column. For example,
+
+```sql
+CREATE TABLE OUTPUT AS SELECT ID, V0, V1, COPY(ID) AS V2 FROM INPUT;
+-- resulting schema: ID INT KEY, V0 INT, V1 INT, V2 INT
+```
+
+### Grouping by multiple expressions.
+
+KsqlDB supports grouping by multiple expressions, for example:
+
+```sql
+SELECT V0, ABS(V1), COUNT(*) AS COUNT FROM INPUT GROUP BY V0, ABS(V1);
+```
+
+However, it does not _yet_ support multiple key columns. If the above is converted to a persistent
+query the key is generated by concatenating the string representation of the grouping expressions.
+For example:
+
+```sql
+CREATE TABLE OUTPUT AS SELECT V0, ABS(V1) AS V1, COUNT(*) AS COUNT FROM INPUT GROUP BY V0, ABS(V1);
+-- resulting schema: KSQL_COL_0 STRING KEY, V0 INT, V1 INT, COUNT BIGINT
+-- where KSQL_COL_0 contains data in the form V0 + "|+|" + ABS(V1)
+```
+
+Even though ksqlDB is currently combining the multiple grouping expressions, we propose that the
+projection should still accept the individual columns, and recognise them as key columns. This will
+be compatible with the upcoming multiple-key-column support.
+
+However, this posses a problem, as it does not provide a single place where the user can provide an
+alias for the system generated `KSQL_COL_0` key column name. Any solution to allow providing an
+alias would likely be incompatible with the planned multiple key column support. 
+
+Hence, we propose leaving this edge case unsolved, i.e. users will _not_ be able to provide an alias
+for the name of the key column resulting from multiple grouping expressions. This will be resolved
+when support for multiple key columns is added.
+
+Alternatively, ksqlDB could support the non-strandard `GROUP BY (a, b) AS c` style aliasing, to allow
```

Comment:
since we're going to deliver multiple keys in the future, I don't see the benefit really of adding something non-standard now